tests/deque.rs

```diff
@@ -0,0 +1,158 @@
+use std::cell::RefCell;
+use std::collections::VecDeque;
+use std::rc::Rc;
+
+use left_right::{
+    aliasing::{Aliased, DropBehavior},
+    Absorb, ReadHandle,
+};
+
+// Value encapsulates an integer value and keeps a registry of live values up to
+// date.
+struct Value {
+    v: i32,
+    r: Rc<ValueRegistry>,
+}
+
+impl Value {
+    fn new(v: i32, r: Rc<ValueRegistry>) -> Self {
+        r.adjust_count(1);
+        Self { v, r }
+    }
+}
+
+impl Drop for Value {
+    fn drop(&mut self) {
+        self.r.adjust_count(-1);
+    }
+}
+
+// ValueRegistry keeps track of the number of Values that have been created and
+// not yet dropped.
+struct ValueRegistry {
+    num_live_values: RefCell<i64>,
+}
+
+impl ValueRegistry {
+    fn new() -> Self {
+        Self {
+            num_live_values: RefCell::new(0),
+        }
+    }
+
+    fn adjust_count(&self, delta: i64) {
+        let mut live_vals = self.num_live_values.borrow_mut();
+        *live_vals += delta;
+        assert!(*live_vals >= 0);
+    }
+
+    fn expect(&self, expected_count: i64) {
+        assert_eq!(*self.num_live_values.borrow(), expected_count);
+    }
+}
+
+struct NoDrop;
```

Comment:
It feels strange that every user has to implement these types. Wouldn't it be much simpler to just use a bool const instead of a type? (`pub struct Aliased<T, const DoDrop: bool>`).

Reply:
I guess the idea is that someone could implement a trait only for `Aliased<T, true>` potentially (in theory) causing the memory layout to change for just the `true` case?

Reply:
The reason why these _have_ to be types private to each implementor is explained in the "Unsafe casting" section [of the `mod aliasing` docs](https://docs.rs/left-right/latest/left_right/aliasing/index.html#unsafe-casting). Basically, it's the only way you can guarantee that the necessary cast is actually sound.